### ISSUE TITLE
Data race during frequent put/get

### ISSUE THREAD

**Issue by karalabe** [syndtr/goleveldb#136]
We've seen some odd crashes in our code, which got worse lately with some performance increases and managed to track it down to a leveldb put/get issue. In essence, during heavy writes and reads, leveldb sometimes reports keys as non existing, even though they were just inserted beforehand. This is very problematic in code which assumes that a written key is immediately available for reading.

The code to reproduce it is:

```go
package main

import (
	"bytes"
	"fmt"
	"io/ioutil"
	"math/big"
	"math/rand"
	"os"
	"runtime"
	"sync"

	"github.com/syndtr/goleveldb/leveldb"
	"github.com/syndtr/goleveldb/leveldb/opt"
	"golang.org/x/crypto/sha3"
)

func main() {
	// Create a temporary leveldb to muck around with
	dir, _ := ioutil.TempDir("", "")
	defer os.RemoveAll(dir)

	db, err := leveldb.OpenFile(dir, &opt.Options{OpenFilesCacheCapacity: 256, BlockCacheCapacity: 256 * opt.MiB, WriteBuffer: 128 * opt.MiB})
	if err != nil {
		panic(err)
	}
	// Start writing on concurrent threads and
	threads := runtime.NumCPU()
	runtime.GOMAXPROCS(threads)

	pend := new(sync.WaitGroup)
	for i := 0; i < threads; i++ {
		pend.Add(1)
		go func(worker int) {
			defer pend.Done()

			// Start writing random junk into the database
			hasher := sha3.New256()
			rander := rand.New(rand.NewSource(int64(worker)))

			for i := 0; i < 1000000; i++ {
				// Generate the next key value pair
				hasher.Write(big.NewInt(rander.Int63()).Bytes())
				key := hasher.Sum(nil)

				hasher.Write(big.NewInt(rander.Int63()).Bytes())
				val := hasher.Sum(nil)

				// Insert into the database and immediately read it back
				if err := db.Put(key, val, nil); err != nil {
					panic(err)
				}
				if back, err := db.Get(key, nil); err != nil {
					panic(err)
				} else if bytes.Compare(val, back) != 0 {
					panic(fmt.Sprintf("save/load mismatch: have %x, want %x", back, val))
				}
				// Output some logs for progress reports
				if i%10000 == 0 {
					fmt.Printf("#%d at %d\n", worker, i)
				}
			}
		}(i)
	}
	pend.Wait()
}
```

Which after about maybe a hal a minute to minute crashes:

```
#4 at 0
[...]
#6 at 320000
panic: leveldb: not found

goroutine 28 [running]:
panic(0x5b18e0, 0xc82000a3d0)
	/opt/google/go/src/runtime/panic.go:464 +0x3e6
main.main.func1(0xc828100310, 0xc820001200, 0x4)
	/work/src/github.com/ethereum/go-ethereum/lvlrace/main.go:55 +0x45b
created by main.main
	/work/src/github.com/ethereum/go-ethereum/lvlrace/main.go:64 +0x1cb
```

The leveldb commit this happens on is latest master

```
commit ad0d8b2ab58a55ed5c58073aa46451d5e1ca1280
Author: Suryandaru Triandana <syndtr@gmail.com>
Date:   Tue Feb 23 08:31:43 2016 +0700

    leveldb: fixes compaction commit racy (fixes #133)
```

---

Edit: In our code, if I add a 1 second sleep to the place where it crashes and refetch the data from leveldb it's actually present, hence why I'm assuming it's a data race.

**Comment by karalabe** [comment]
Btw, if you run the above code with the race detector, it almost instantly produces a crash:

```
$ go run --race main.go 
#0 at 0
#1 at 0
#2 at 0
#3 at 0
#4 at 0
#5 at 0
#7 at 0
#6 at 0
panic: leveldb: not found

goroutine 68 [running]:
panic(0x64fca0, 0xc82000a3c0)
	/opt/google/go/src/runtime/panic.go:464 +0x3ff
main.main.func1(0xc828110300, 0xc820001380, 0x4)
	/work/src/github.com/ethereum/go-ethereum/lvlrace/main.go:54 +0x4e9
created by main.main
	/work/src/github.com/ethereum/go-ethereum/lvlrace/main.go:63 +0x23c
exit status 2
```

**Comment by karalabe** [comment]
Gotcha, it's a logical race with acking the pending writes. Will open a PR in a moment :)

**Comment by karalabe** [comment]
Closing this as fixed by #137 :)
